### PR TITLE
Add basic meteor swarm

### PR DESCRIPTION
**Pull request by metalgearsloth**
Fixes https://github.com/space-wizards/space-station-14/issues/4316

:cl:
- add: Add meteor swarm event.

### CODE REVIEW

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                var offset = angle.RotateVec(new Vector2((maximumDistance - minimumDistance) * robustRandom.NextFloat() + minimumDistance, 0));
+                var spawnPosition = new MapCoordinates(center + offset, mapId);
+                var meteor = entityManager.SpawnEntity("MeteorLarge", spawnPosition);
+                var physics = compManager.GetComponent<PhysicsComponent>(meteor.Uid);
+                physics.BodyStatus = BodyStatus.InAir;
+                physics.ApplyLinearImpulse(-offset.Normalized * MeteorVelocity * physics.Mass);
+                physics.ApplyAngularImpulse(
+                    // Get a random angular velocity.
+                    physics.Mass * ((MaxAngularVelocity - MinAngularVelocity) * robustRandom.NextFloat() +
+                                    MinAngularVelocity));
```
Comment (mirrorcult) [inline Content.Server/StationEvents/Events/MeteorSwarm.cs:120]: feels like there could be a helper for this, i imagine a lot of stuff would want to do something similar but out of scope

**Comment by ike709** [comment]
Is this the best idea when the atmospherics department isn't done and able to deal with breaches?

**Comment by Ygg01** [comment]
> Is this the best idea when the atmospherics department isn't done and able to deal with breaches?

Atm. Meteor deals too little damage to cause a breach.

**Comment by PaulRitter** [comment]
> Is this the best idea when the atmospherics department isn't done and able to deal with breaches?

here at spacewizards we've decided to innovate onto the idea of test-driven-development, introducing crisis-driven-development

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+        public override void Update(float frameTime)
+        {
+            base.Update(frameTime);
+
+            if (!Started) return;
+
+            if (_waveCounter <= 0)
+            {
+                EndAfter = float.MinValue;
+            }
```
Comment (PaulRitter) [inline Content.Server/StationEvents/Events/MeteorSwarm.cs:73]: this seems to be missing a return statement?

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+            if (_cooldown > 0f) return;
+
+            _waveCounter--;
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+
+            _cooldown += (MaximumCooldown - MinimumCooldown) * robustRandom.NextFloat() + MinimumCooldown;
+            var entityManager = IoCManager.Resolve<IEntityManager>();
+            var mapManager = IoCManager.Resolve<IMapManager>();
+            var compManager = IoCManager.Resolve<IComponentManager>();
```
[the remark this answers is not in the document]
Comment (PaulRitter) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:84]: yes, cache these

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+                EndAfter = float.MinValue;
+            }
+            _cooldown -= frameTime;
+
+            if (_cooldown > 0f) return;
+
+            _waveCounter--;
+            var robustRandom = IoCManager.Resolve<IRobustRandom>();
+
+            _cooldown += (MaximumCooldown - MinimumCooldown) * robustRandom.NextFloat() + MinimumCooldown;
```
Comment (PaulRitter) [inline Content.Server/StationEvents/Events/MeteorSwarm.cs:81]: add an extension-helper to robustrandom whenst (i dont want you to in this pr, im just saying it'd probably be something nice to have)

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+
+            Box2? playableArea = null;
+            var mapId = EntitySystem.Get<GameTicker>().DefaultMap;
+
+            foreach (var grid in mapManager.GetAllGrids())
+            {
+                if (grid.ParentMapId != mapId || !compManager.TryGetComponent(grid.GridEntityId, out PhysicsComponent? gridBody)) continue;
+                var aabb = gridBody.GetWorldAABB();
+                playableArea ??= aabb;
+                playableArea = playableArea.Value.Union(aabb);
```
Comment (PaulRitter) [inline Content.Server/StationEvents/Events/MeteorSwarm.cs:94]: what is going on here. What does union do? not saying its wrong but my codin' sense is tingling

File: Content.Server/StationEvents/Events/MeteorSwarm.cs [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+            }
```
Comment (PaulRitter) [inline Content.Server/StationEvents/Events/MeteorSwarm.cs:95]: come to think of it, this could probably also be done in the init?

**Review by PaulRitter — CHANGES_REQUESTED** [review]
god we need robustrandom helpers

good code, fun thing, proud of you son👍

**Comment by mirrorcult** [comment]
> 
> 
> Is this the best idea when the atmospherics department isn't done and able to deal with breaches?

oxygen/air canisters can be bought and walls can be rebuilt currently, so seems fine

**Comment by SweptWasTaken** [comment]
> 
> 
> > Is this the best idea when the atmospherics department isn't done and able to deal with breaches?
> 
> oxygen/air canisters can be bought and walls can be rebuilt currently, so seems fine

There's also a radiation storm event but we don't have functional rad-suits so technically the station is better equipped to deal with a meteor storm than a rad-storm.

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 2 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:73]: I must've accidentally nuked it when pushing

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 5 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:94]: It gets the minimum box that covers both of them. It's also what physics AABB uses as well.

**Comment by metalgearsloth** [comment]
Events don't run normally and the explosions are extremely weak atm. Sometimes you dont even get a breach at all.

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 5 above)
Comment (PaulRitter) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:94]: i dont know how expensive union is but wouldn't it be better to
```suggestion
                playableArea = playableArea == null ? aabb : playableArea.Value.Union(aabb);
```

to cut the unneccesary union call when playableArea is null? I dont know how often it'd happen.

Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:94]: It's just the Box2 equivalent of doing a Min / Max call, and given it's happening once every n seconds it won't even show on the profiler but sure.

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 6 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:95]: Reason for that is in case there's moving grids to avoid spawning on top of them.

Comment (PaulRitter) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs:95]: a right forgot about that, sorry

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 1 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs]: Made an issue for it.

File: Content.Server/StationEvents/Events/MeteorSwarm.cs (hunk 4 above)
Comment (metalgearsloth) [reply Content.Server/StationEvents/Events/MeteorSwarm.cs]: Made an issue on robust for it.

**Pull request merged**